pkg/translate24to31/translate24to31.go

```diff
@@ -0,0 +1,443 @@
+// Copyright 2019 Red Hat, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package translate24to31
+
+import (
+	"errors"
+	"fmt"
+	"path/filepath"
+	"reflect"
+
+	old "github.com/coreos/ignition/config/v2_4/types"
+	oldValidate "github.com/coreos/ignition/config/validate"
+	"github.com/coreos/ignition/v2/config/v3_1/types"
+	"github.com/coreos/ignition/v2/config/validate"
+
+	"github.com/coreos/ign-converter/pkg/util"
+)
+
+// Check2_4 returns if the config is translatable but does not do any translation.
+// fsMap is a map from v2 filesystem names to the paths under which they should
+// be mounted in v3.
+func Check2_4(cfg old.Config, fsMap map[string]string) error {
+	rpt := oldValidate.ValidateWithoutSource(reflect.ValueOf(cfg))
+	if rpt.IsFatal() || rpt.IsDeprecated() {
+		// disallow any deprecated fields
+		return fmt.Errorf("Invalid input config:\n%s", rpt.String())
+	}
+
+	if len(cfg.Networkd.Units) != 0 {
+		return util.UsesNetworkdError
+	}
+
+	// check that all filesystems have a path
+	if fsMap == nil {
+		fsMap = map[string]string{}
+	}
+	fsMap["root"] = "/"
+	for _, fs := range cfg.Storage.Filesystems {
+		if _, ok := fsMap[fs.Name]; !ok {
+			return util.NoFilesystemError(fs.Name)
+		}
+	}
+
+	// check that there are no duplicates with files, links, or directories
+	// from path to a pretty-printing description of the entry
+	entryMap := map[string]string{}
+	links := make([]string, 0, len(cfg.Storage.Links))
+	// build up a list of all the links we write. We're not allow to use links
+	// that we write
+	for _, link := range cfg.Storage.Links {
+		path := filepath.Join("/", fsMap[link.Filesystem], link.Path)
+		links = append(links, path)
+	}
+
+	for _, file := range cfg.Storage.Files {
+		path := filepath.Join("/", fsMap[file.Filesystem], file.Path)
+		name := fmt.Sprintf("File: %s", path)
+		if duplicate, isDup := entryMap[path]; isDup {
+			return util.DuplicateInodeError{duplicate, name}
```

Comment:
23to30 returns a pointer to this, noisifying the diff.

Reply:
I don't follow. Which part to you mean specifically?

Reply:
Sorry, I put this comment on the wrong line.  I was referring to https://github.com/coreos/ign-converter/blob/8bfca335dd8aeaf28b5dec8fd6f4cf87bfe06786/translate/v23tov30/v23tov30.go#L99.